director/sort_algorithms.go

```diff
@@ -57,29 +58,83 @@ func distanceOnSphere(lat1 float64, long1 float64, lat2 float64, long2 float64)
-	return arc / math.Pi
+	return arc
 }
 
 // Create a weight between [0,1] that indicates a priority. The returned weight is directly correlated
 // with priority (higher weight is higher priority is lower distance)
-func distanceWeight(coord Coordinate, ad server_structs.ServerAd) float64 {
-	return 1 - distanceOnSphere(coord.Lat, coord.Long, ad.Latitude, ad.Longitude)
+//
+// Is realD set to true, then return the distance between the two coordinates in miles
+func distanceWeight(lat1 float64, long1 float64, lat2 float64, long2 float64, realD bool) float64 {
+	if realD {
+		return distanceOnSphere(lat1, long1, lat2, long2) * earthRadiusToMilesFactor
+	} else {
+		return 1 - (distanceOnSphere(lat1, long1, lat2, long2) / math.Pi)
+	}
 }
 
-// Create a weight between [0,1] that indicates a priority. The returned weight is directly correlated
-// with priority (higher weight is higher priority)
-func distanceAndLoadWeight(coord Coordinate, sAd server_structs.ServerAd) float64 {
-	distance := distanceOnSphere(coord.Lat, coord.Long, sAd.Latitude, sAd.Longitude)
+// Given the input value, return a weight [0, 1.0] based on the gated havling of the base weight 1.0.
+//   - If the input value is between 0.0 and the threshold, return 1.0.
+//   - If the input value is above the threshold, the weight decreases by half for every halvingFactor units of the input value
+func gatedHavlingMultiplier(val float64, threshold float64, halvingFactor float64) float64 {
+	if halvingFactor == 0 || threshold == 0 {
+		return 1.0
+	}
+	if val >= 0.0 && val <= threshold {
+		return 1.0
+	} else {
+		// multiplier decreases by half for every havlvingFactor units of the value for value above the threshold
+		base := math.Max(1, (math.Floor((val-threshold)/halvingFactor))*2)
+		multiplier := 1.0 / base
+		return multiplier
+	}
+}
 
-	// For now, load is always 0.5. Eventually we'll pull this value from the server ad. Furthermore,
-	// assume a server's load has more impact on its performance than its distance does. As long as load is
-	// hard coded, this function should act exactly like distanceWeight.
-	// TODO: Come up with a better function once we have an actual load value and know how it works
-	load := 0.5
+// Given a SwapMaps struct, stochasticlly sort the weights based on the folliwng procedure:
+//
+//  1. Create ranges [0, weight_1), [weight_1, weight_1 + weight_2), ... for each weight.
+//
+//  2. Select a random number in the range [0, sum(weights)).
+//
+//  3. If the number falls within the range corresponding to the weight, it is sorted to the top.
+//
+//  4. Repeat step 2-3 to select a the rest weights
+//
+// Returnss the sorted list of SwapMap.Index and the generated random weights for reference.
+// You may specify the maxOut argument to limit the output.
+func stochasticSort(sm SwapMaps, maxOut int) (candidates []int, randWeights []float64) {
+	if maxOut == 0 {
+		maxOut = len(sm)
+	} else if maxOut > len(sm) {
+		return
```

Comment:
Why return here? If we tell the sort algorithm "don't give me more than 10 back" but there are only 5 candidates, shouldn't we still sort those 5 and return them?

Reply:
See reply above